README.rst
=======================
BRAin NEtwork STAbility
=======================


.. image:: https://data.caltech.edu/badge/doi/10.1073/pnas.1913042117.svg
        :target: https://doi.org/10.1073/pnas.1913042117

.. image:: https://img.shields.io/pypi/v/branesta.svg
        :target: https://pypi.python.org/pypi/branesta

.. image:: https://img.shields.io/travis/BotondA/branesta.svg
        :target: https://travis-ci.org/BotondA/branesta

.. image:: https://readthedocs.org/projects/branesta/badge/?version=latest
        :target: https://branesta.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://img.shields.io/badge/License-MIT-brightgreen.svg
     :target: https://opensource.org/licenses/MIT

| 

.. image:: https://raw.githubusercontent.com/BotondA/branesta/v0.1.6/assets/logo_w_text.png


Introduction
------------

Branesta is a tool for computing brain network stability, a biomarker for brain aging.

* Free software: MIT license
* Documentation: https://branesta.readthedocs.io.

Please cite our article:

Mujica-Parodi, Lilianne R., et al. "Diet modulates brain network stability, a biomarker for brain aging, in young adults." Proceedings of the National Academy of Sciences 117.11 (2020): 6170-6177.
link: https://www.pnas.org/content/117/11/6170


Description
------------

Brain network stability measures the extent of temporal reorganization that takes place in brain networks. Brain networks describe inter-regional communication across the brain. Lower network stability (represented by higher values) is related to weaker persistence of brain networks. The terms Network Stability and Network INstability are used interchangibly and they refer to the exact same metric.

The procedure of computing brain network stability is as follows: fMRI time-series that were previously parcelled into ROIs are first binned into time windows (=snapshots) of N timepoints without overlaps (N = window length). Next, pairwise correlations among all ROIs are computed separately for each time window. For the whole brain, brain network stability (scalar) is quantified by taking the l2 norm of the element-wise differences of correlation matrices corresponding to two different snapshots. τ is the number of steps separating two snapshots from which a given value of brain network stability is calculated from. For instance, if τ=1, two consecutive snapshots snapshots are used (e.g. #4 and #5). If τ=16, then 16 snapshots are separating the two snapshots (e.g. #3 and #19). Given a window length of 30 timepoints, if the time-series have a length of 720 timepoints, then there will be 24 snapshots (720/30=24). At τ=1, there are 23 instability values, whereas at τ=20, 4 different instability values would be computed.

For functional networks (labeled as "subnetworks" in our program), the procedure is analog to the above. The only difference is that once correlations are computed for each time window, element-wise differences are calculated only across those ROIs that spatially overlap with the functional network. In order to facilitate comparison of network instability among networks consisting of different number of nodes, network stability is normalized with the number of edges in the correlation matrix. 

Features
--------

* computes network stability from parcelled time-series
* performs computations at every τ
* computes for subnetworks (optional)
* allows user-defined time window length
* easy to install (pip)
* command line tool

Credits
-------

This package was developed within the Laboratory for Computational Neurodiagnostics (LCNeuro_) at Stony Brook University, New York.

.. _LCNeuro: https://lcneuro.org

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.9 (2021-4-16)
------------------

* bug fixes


0.1.8 (2021-4-15)
------------------

* code revisions, improved readability


0.1.7 (2020-12-22)
------------------

* Logo fix


0.1.6 (2020-12-22)
------------------

* Added logo


0.1.5 (2020-12-22)
------------------

* Added unittests.
* Updated imports.


0.1.4 (2020-11-29)
------------------

* Updated logging.


0.1.3 (2020-11-28)
------------------

* First standalone release.


0.1.2 (2020-11-28)
------------------

* Updated dependencies.


0.1.1 (2020-11-27)
------------------

* Added analysis module.


0.1.0 (2020-11-26)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for branesta."""


tests/test_branesta.py
#!/usr/bin/env python

"""Tests for `branesta` package."""

import os
import glob
import unittest
import pandas as pd

from branesta import branesta


class TestBranesta(unittest.TestCase):
    """Tests for `branesta` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

        # Run branesta, output results into temp dir
        branesta.analyze(
                "tests", "tests/testdata/ts", "tests/temp/", win_len=30,
                subnetpath="tests/testdata/subnets/subnetworks_willard.csv",
                tot_len=720, tot_roi_num=498
            )

    def tearDown(self):
        """Tear down test fixtures, if any."""

        # Delete temp outputs
        os.system("rm tests/temp/*")


    def test_branesta(self):
        """Test whether output csv with stability values is as expected."""

        # Open output
        out = pd.read_csv("tests/temp/brain_network_stability_tests.csv")

        # Open reference
        ref = pd.read_csv(glob.glob("tests/ref/brain_network_stability_*.csv")[-1])

        # Print some info
        print("Info from test-case:", "\nshape:", out.shape, "\nhead:\n", out.head())

        # Assert equality
        pd.testing.assert_frame_equal(out, ref)



branesta/cli.py
"""Console script for branesta."""

import sys
import argparse
from datetime import datetime
from branesta import branesta

def main():
    """Console script for branesta."""
    # Unique identifier for current analysis
    ver = datetime.now().strftime("%Y%m%d%H%M%S")

    # Parse user inputs
    parser = argparse.ArgumentParser(description="Arguments for analysis.")
    parser.add_argument("srcdir", default="./",
                        help="path to source directory with time-series")
    parser.add_argument("outdir", default="./",
                        help="path to out directory for results and logs")
    parser.add_argument('win_len', default=30, type=int,
                        help="number of time-frames in each snapshot")
    parser.add_argument('-s', '--subnetpath',
                        help="path to csv file with subnetwork labels and ixs, optional")
    parser.add_argument('-l', '--tot_len', type=int,
                        help="expected total number of time frames, optional")
    parser.add_argument('-r', '--tot_roi_num', type=int,
                        help="expected total number of ROIs, optional")
    args = parser.parse_args()

    # Call analysis
    branesta.analyze(ver, **vars(args))

    return 0


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


branesta/branesta.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun May 24 01:01:19 2020

@author: botond antal

Notes:
    This application computes network stability from a set of time-series.
"""

import os
import argparse
import logging
import numpy as np
import pandas as pd
import itertools
from nilearn.connectome import ConnectivityMeasure


class NetworkMetricsPipeline:
    """
    Class with methods for calculating network stability
    from ROI space time-series
    """

    def __init__(self, roi_time_series):

        # Convert input time_series to numpy array
        self.roi_time_series = np.array(roi_time_series)
        self.t_len = roi_time_series.shape[0]

    def calc_windows(self, win_len):
        """
        Divides up time-series into time windows.
        Outputs array of windows.
        """

        # Number of windows-1
        num = int((self.t_len - 1 - (self.t_len - 1) % win_len) / win_len)

        # Indicies at which slicing should occur
        indicies = np.linspace(win_len, num*win_len, num).astype(int)

        # 30 frame long time windows (last window might be shorter)
        windows = np.split(self.roi_time_series, indicies, axis=0)

        # Remove last window if shorter than window_size
        windows = windows[:-1] if self.t_len % win_len > 0 else windows

        # Given that all dimensions are consistent, convert list to numpy array
        self.windows = np.array(windows)

        return self

    def calc_corrs(self):
        """
        Calculates correlations between ROIs.
        Outputs a list of correlation matrices corresponding to windows.
        """

        # Take pairwise correlation of ROI tracers separately within all windows
        correlation_measure = ConnectivityMeasure(kind='correlation')
        self.corr_matrices = correlation_measure.fit_transform(self.windows)

        return self

    def calc_stab(self, subnet_ixs):
        """
        Calculates network stability.
        Outputs a (list of) list of network stability values.
        One value for each subnetwork within one row for each tau.
        """

        # Array of tau values
        self.tau_vals = np.arange(1, len(self.corr_matrices))

        # Array for storing normalized stability values
        self.stability = []

        # Iterate over all taus
        for i, tau in enumerate(self.tau_vals):

            # Pairs of windows at a given tau which the element-wise
            # differences of correlations will be calculated between
            window_pairs = [[start, start + tau] \
                       for start in range(len(self.corr_matrices) - tau)]

            # Calculate elementwise differences
            corr_diffs_ls = [
                self.corr_matrices[win1] - self.corr_matrices[win2] \
                                 for win1, win2 in window_pairs]

            # Concatenate elementwise differences
            corr_diffs = np.array(corr_diffs_ls)

            # Preallocate for temporarily storing stability values,
            # filled with nans
            subnets_stab = np.full((len(subnet_ixs), len(self.tau_vals)), np.nan)

            # Iterate over all fROIs
            for j, ixs in enumerate(subnet_ixs.values()):

                # Extract corr_diffs corresponding to fROI
                subnet_corr_diffs = corr_diffs[:, ixs[0], ixs[1]]

                # Take l2norm
                stabs_prenorm = np.linalg.norm(subnet_corr_diffs, axis=1)

                # Normalize with sqroot of number of non-main-diagonal edges
                subnets_stab[j][:len(corr_diffs)] = stabs_prenorm/ \
                    np.sqrt(np.sqrt(len(ixs.T))*(np.sqrt(len(ixs.T))-1))

            # Store computed stability values
            self.stability.append(list(subnets_stab.T))

        return self

    def convert_stab_to_df(self, meta, subnet_ixs):
        """
        Converts computed stability values to dataframe.
        Outputs pandas dataframe with network stability values
        and labeled columns.
        """

        # Indexes
        mtx = pd.MultiIndex.from_frame(
                 pd.DataFrame(
                         itertools.product(
                                 self.tau_vals,
                                 self.tau_vals,
                                 list(subnet_ixs.keys()))) \
                    .T.set_index(pd.Index(["tau", "time", "subnetwork"])).T \
                    .assign(**meta) \
                    .loc[:, list(meta.keys()) + ["tau", "time", "subnetwork"]])

        # Construct dataframe of stabilitities for current instance
        self.df = pd \
            .DataFrame(self.stability) \
            .T.melt() \
            .drop(labels="variable", axis=1) \
            .explode(column="value") \
            .set_index(mtx) \
            .rename({"value": "stability"}, axis=1)

        return self


def compute_network_stability(
        file, meta={}, srcdir=None, win_len=30, subnet_ixs={},
        tot_roi_num=None, tot_len=None):

    """
    Wrapper function for calculating stability values for a single time-series
    entity which was already preprocessed and converted into ROI space

    Parameters
    ----------
    file: str
        name of file containing time-series, located in srcdir, must follow BIDS
        naming convention. Shape must be: [samples, channels] or [time, space]
    srcdir: str
        path to source directory with containing data files
    win_len: int
        number of frames in time windows/snapshots considered for computing
        network stability
    subnet_ixs: dict
        dictionary of indexes specific to functional ROIs, the format is
        {key=func_label of the fROI, value=2D array of the product of
        ROI indexes included in the fROI},
        examples for labels: "whole_brain", "visuospatial", "auditory"
    tot_roi_num: int
        optional, total number of ROIs expected in time-series
    tot_len: int
        optional, total number of time frames expected in time-series
    """

    # Load data
    # --------

    # Status
    # text = f"Loading {file}"
#    print(text)
    # logging.info(text)

    # Construct meta data according to BIDS
    # example: {'sub': '057', 'ses': 'glc', 'task': 'task', 'run': '2'}
    file_info = file.split(".")[0].split("_")
    meta = {}

    for item in file_info:
        id_, info = item.split("-")
        meta[id_] = info

    # Load ROI space time-series
    roi_time_series = pd.read_csv(os.path.join(srcdir, file),
                                  header=None, index_col=False)
    roi_time_series = np.array(roi_time_series)

    # Add subnetwork with all ROIs called whole brain
    subnet_ixs["whole_brain"] = \
        np.array(list(itertools.product(
                np.arange(roi_time_series.shape[1]),
                np.arange(roi_time_series.shape[1])))).T

    # Check shape of time-series if expected values are provided
    if tot_len and tot_roi_num:
        if roi_time_series.shape != (tot_len, tot_roi_num):
            text = f"\n[!] Abnormal shape {roi_time_series.shape} in item: " \
            f"{file}! Item will be excluded!\n"
#            print(text)
            logging.warning(text)
            return

    # Check for zero/close-to-zero ROIs
    roi_means =  np.mean(np.abs(roi_time_series), axis=0)
    zero_rois = np.argwhere(roi_means < 1e-10)
    if len(zero_rois) > 0:
        text = f"[!] Zero ROI(s) {zero_rois} found in item: {file}! Proceeding."
#        print(text)
        logging.warning(text)

    # Compute metrics
    # -------

    # Status
    text = f"Computing metrics for {file}"
#    print(text)
    logging.info(text)

    # Compute stability
    df = NetworkMetricsPipeline(roi_time_series) \
        .calc_windows(win_len) \
        .calc_corrs() \
        .calc_stab(subnet_ixs) \
        .convert_stab_to_df(meta, subnet_ixs) \
        .df

    # Status
    # text = f"Finished computations for {file}"
#    print(text)
    # logging.info(text)
#
    # Returns
    return df


def analyze(ver, srcdir, outdir, subnetpath=None, **opts):
    """
    Computes network stability for files in srcdir. Considers only
    .csv files. Naming must follow BIDS format.

    Parameters
    ----------
    ver: str
        version of current analysis
    srcdir: str
        path to directory with data files to analyze
    outdir: dict
        path where results are dumped
    subnetpath: str
        path to csv with subnetwork labels and indexes
    opts: dict
        dictionary with optional keyword arguments for computations
    """

    # Set up logger
    logging.basicConfig(level=logging.INFO,
                        format='%(asctime)s %(levelname)-8s %(message)s',
                        handlers=[
                                logging.FileHandler(
                                    filename=os.path.join(outdir, f"{ver}.log")),
                                logging.StreamHandler()],
                        datefmt='%Y-%m-%d %H:%M:%S')

    # Dict to store edge indexes for each subnetwork
    subnet_ixs = {}

    # If path was given, fill up subnet labels and ixs
    if subnetpath:

        # Import subnetwork labels and ROI indexes
        subnet_data = pd.read_csv(subnetpath)

        # Get subnet labels
        subnet_labels = subnet_data["network"].unique()

        # Iterate over all subnetworks
        for label in subnet_labels:

            # Extract ROI indexes found within fROI
            ids = subnet_data \
                .loc[subnet_data["network"] == label, "roi"].to_numpy() - 1

            # Assign combinations of ROI indexes representing edges within subnet
            subnet_ixs[f"{label}"] = np.array(list(itertools.product(ids, ids))).T


    # Build analysis_kwargs dict
    analysis_kwargs = {"srcdir": srcdir, "subnet_ixs": subnet_ixs, **opts}

    # Status
    text = 'Starting analysis with the following settings:\n' \
           + "\n".join([": ".join([str(key), str(value)]) \
                        if key != "subnet_ixs" else  \
                        ": ".join([str(key), str(list(value.keys()))]) \
                        for key, value in analysis_kwargs.items()])
#    print(text)
    logging.info(text)

    # Collect files to analyze
    files = sorted([file for file in os.listdir(srcdir) if ".csv" in file])

    # Initiate collection of outputs
    out = []

    # Analyze files
    for file in files:

        # Compute metrics
        res = compute_network_stability(
                file=file,
                **analysis_kwargs
                )

        # If file was not skipped and output exists
        if type(res) == pd.DataFrame:

            # Append results to collection of outputs
            out.append(res)

    # Convert outputs pandas dataframe and write into csv
    pd.concat(out, axis=0).to_csv(os.path.join(outdir, f"brain_network_stability_{ver}.csv"))

    # Status
    text = f"Analysis has finished. Number of files analyzed: {len(out)}."
#    print(text)
    logging.info(text)

    # Shut down logging
    logging.shutdown()


